{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "if os.getcwd().endswith('notebook'):\n",
    "    os.chdir('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.set(palette='colorblind', font_scale=1.3)\n",
    "palette = sns.color_palette()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load assemblies metadata file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "      <th>strain</th>\n",
       "      <th>assembly_level</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_accession</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GCA_000010525.1</th>\n",
       "      <td>438753</td>\n",
       "      <td>7</td>\n",
       "      <td>Azorhizobium caulinodans ORS 571</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhizobiales</td>\n",
       "      <td>Xanthobacteraceae</td>\n",
       "      <td>Azorhizobium</td>\n",
       "      <td>Azorhizobium caulinodans</td>\n",
       "      <td>Azorhizobium caulinodans ORS 571</td>\n",
       "      <td>Complete Genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000218545.1</th>\n",
       "      <td>593907</td>\n",
       "      <td>11</td>\n",
       "      <td>Cellulomonas gilvus ATCC 13127</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Actinobacteria</td>\n",
       "      <td>Micrococcales</td>\n",
       "      <td>Cellulomonadaceae</td>\n",
       "      <td>Cellulomonas</td>\n",
       "      <td>Cellulomonas gilvus</td>\n",
       "      <td>Cellulomonas gilvus ATCC 13127</td>\n",
       "      <td>Complete Genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000020965.1</th>\n",
       "      <td>309799</td>\n",
       "      <td>14</td>\n",
       "      <td>Dictyoglomus thermophilum H-6-12</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Dictyoglomi</td>\n",
       "      <td>Dictyoglomia</td>\n",
       "      <td>Dictyoglomales</td>\n",
       "      <td>Dictyoglomaceae</td>\n",
       "      <td>Dictyoglomus</td>\n",
       "      <td>Dictyoglomus thermophilum</td>\n",
       "      <td>Dictyoglomus thermophilum H-6-12</td>\n",
       "      <td>Complete Genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_000012885.1</th>\n",
       "      <td>338963</td>\n",
       "      <td>19</td>\n",
       "      <td>Pelobacter carbinolicus DSM 2380</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Deltaproteobacteria</td>\n",
       "      <td>Desulfuromonadales</td>\n",
       "      <td>Desulfuromonadaceae</td>\n",
       "      <td>Pelobacter</td>\n",
       "      <td>Pelobacter carbinolicus</td>\n",
       "      <td>Pelobacter carbinolicus DSM 2380</td>\n",
       "      <td>Complete Genome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GCA_001735525.1</th>\n",
       "      <td>23</td>\n",
       "      <td>23</td>\n",
       "      <td>Shewanella colwelliana</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Alteromonadales</td>\n",
       "      <td>Shewanellaceae</td>\n",
       "      <td>Shewanella</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Shewanella colwelliana</td>\n",
       "      <td>Scaffold</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     taxid  species_taxid                     organism_name  \\\n",
       "assembly_accession                                                            \n",
       "GCA_000010525.1     438753              7  Azorhizobium caulinodans ORS 571   \n",
       "GCA_000218545.1     593907             11    Cellulomonas gilvus ATCC 13127   \n",
       "GCA_000020965.1     309799             14  Dictyoglomus thermophilum H-6-12   \n",
       "GCA_000012885.1     338963             19  Pelobacter carbinolicus DSM 2380   \n",
       "GCA_001735525.1         23             23            Shewanella colwelliana   \n",
       "\n",
       "                      domain          phylum                class  \\\n",
       "assembly_accession                                                  \n",
       "GCA_000010525.1     Bacteria  Proteobacteria  Alphaproteobacteria   \n",
       "GCA_000218545.1     Bacteria  Actinobacteria       Actinobacteria   \n",
       "GCA_000020965.1     Bacteria     Dictyoglomi         Dictyoglomia   \n",
       "GCA_000012885.1     Bacteria  Proteobacteria  Deltaproteobacteria   \n",
       "GCA_001735525.1     Bacteria  Proteobacteria  Gammaproteobacteria   \n",
       "\n",
       "                                 order               family         genus  \\\n",
       "assembly_accession                                                          \n",
       "GCA_000010525.1            Rhizobiales    Xanthobacteraceae  Azorhizobium   \n",
       "GCA_000218545.1          Micrococcales    Cellulomonadaceae  Cellulomonas   \n",
       "GCA_000020965.1         Dictyoglomales      Dictyoglomaceae  Dictyoglomus   \n",
       "GCA_000012885.1     Desulfuromonadales  Desulfuromonadaceae    Pelobacter   \n",
       "GCA_001735525.1        Alteromonadales       Shewanellaceae    Shewanella   \n",
       "\n",
       "                                      species  \\\n",
       "assembly_accession                              \n",
       "GCA_000010525.1      Azorhizobium caulinodans   \n",
       "GCA_000218545.1           Cellulomonas gilvus   \n",
       "GCA_000020965.1     Dictyoglomus thermophilum   \n",
       "GCA_000012885.1       Pelobacter carbinolicus   \n",
       "GCA_001735525.1                           NaN   \n",
       "\n",
       "                                              strain   assembly_level  \n",
       "assembly_accession                                                     \n",
       "GCA_000010525.1     Azorhizobium caulinodans ORS 571  Complete Genome  \n",
       "GCA_000218545.1       Cellulomonas gilvus ATCC 13127  Complete Genome  \n",
       "GCA_000020965.1     Dictyoglomus thermophilum H-6-12  Complete Genome  \n",
       "GCA_000012885.1     Pelobacter carbinolicus DSM 2380  Complete Genome  \n",
       "GCA_001735525.1               Shewanella colwelliana         Scaffold  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblies = pd.read_csv(os.path.join(os.getcwd(), 'data/assemblies.csv'), index_col='assembly_accession')\n",
    "assemblies.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Statistics\n",
    "\n",
    "### Number of unique species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of assemblies: 19,741\n",
      "Number of unique species: 19,228\n"
     ]
    }
   ],
   "source": [
    "n_unique_species = len(assemblies['species_taxid'].unique())\n",
    "print(f'Number of assemblies: {len(assemblies):,}')\n",
    "print(f'Number of unique species: {n_unique_species:,}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bacteria vs Archaea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>domain</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Archaea</th>\n",
       "      <td>1419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria</th>\n",
       "      <td>18322</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          count\n",
       "domain         \n",
       "Archaea    1419\n",
       "Bacteria  18322"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "domain_count = assemblies[['domain', 'taxid']].groupby('domain').count()\n",
    "domain_count.columns = ['count']\n",
    "domain_count"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Taxonomy: phylum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 20 phylum (including N/A), representing 92% of all assemblies:\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>percent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Bacteria</th>\n",
       "      <th>Proteobacteria</th>\n",
       "      <td>6755</td>\n",
       "      <td>34.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Firmicutes</th>\n",
       "      <td>3635</td>\n",
       "      <td>18.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidetes</th>\n",
       "      <td>2302</td>\n",
       "      <td>11.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Actinobacteria</th>\n",
       "      <td>1885</td>\n",
       "      <td>9.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea</th>\n",
       "      <th>Euryarchaeota</th>\n",
       "      <td>999</td>\n",
       "      <td>5.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">Bacteria</th>\n",
       "      <th>Cyanobacteria</th>\n",
       "      <td>471</td>\n",
       "      <td>2.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N/A</th>\n",
       "      <td>371</td>\n",
       "      <td>1.9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chloroflexi</th>\n",
       "      <td>317</td>\n",
       "      <td>1.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Spirochaetes</th>\n",
       "      <td>225</td>\n",
       "      <td>1.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Verrucomicrobia</th>\n",
       "      <td>195</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tenericutes</th>\n",
       "      <td>157</td>\n",
       "      <td>0.8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Planctomycetes</th>\n",
       "      <td>127</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea</th>\n",
       "      <th>Crenarchaeota</th>\n",
       "      <td>113</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria</th>\n",
       "      <th>Acidobacteria</th>\n",
       "      <td>109</td>\n",
       "      <td>0.6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea</th>\n",
       "      <th>Thaumarchaeota</th>\n",
       "      <td>96</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteria</th>\n",
       "      <th>Nitrospirae</th>\n",
       "      <td>90</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Archaea</th>\n",
       "      <th>N/A</th>\n",
       "      <td>75</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">Bacteria</th>\n",
       "      <th>Deinococcus-Thermus</th>\n",
       "      <td>75</td>\n",
       "      <td>0.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Candidatus Omnitrophica</th>\n",
       "      <td>68</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Elusimicrobia</th>\n",
       "      <td>63</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  count  percent\n",
       "domain   phylum                                 \n",
       "Bacteria Proteobacteria            6755     34.2\n",
       "         Firmicutes                3635     18.4\n",
       "         Bacteroidetes             2302     11.7\n",
       "         Actinobacteria            1885      9.5\n",
       "Archaea  Euryarchaeota              999      5.1\n",
       "Bacteria Cyanobacteria              471      2.4\n",
       "         N/A                        371      1.9\n",
       "         Chloroflexi                317      1.6\n",
       "         Spirochaetes               225      1.1\n",
       "         Verrucomicrobia            195      1.0\n",
       "         Tenericutes                157      0.8\n",
       "         Planctomycetes             127      0.6\n",
       "Archaea  Crenarchaeota              113      0.6\n",
       "Bacteria Acidobacteria              109      0.6\n",
       "Archaea  Thaumarchaeota              96      0.5\n",
       "Bacteria Nitrospirae                 90      0.5\n",
       "Archaea  N/A                         75      0.4\n",
       "Bacteria Deinococcus-Thermus         75      0.4\n",
       "         Candidatus Omnitrophica     68      0.3\n",
       "         Elusimicrobia               63      0.3"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblies_tax = assemblies.copy()\n",
    "assemblies_tax['phylum'] = assemblies_tax['phylum'].fillna('N/A')\n",
    "phylum_count = assemblies_tax[['domain', 'phylum', 'taxid']].groupby(['domain', 'phylum']).count()\n",
    "phylum_count.columns = ['count']\n",
    "phylum_count['percent'] = phylum_count['count'].apply(lambda c: np.round(100 * c / len(assemblies), 1))\n",
    "\n",
    "top_k = 20\n",
    "top_k_phylum = phylum_count.sort_values('count', ascending=False).head(top_k)\n",
    "print(f\"Top {top_k} phylum (including N/A), representing {top_k_phylum['percent'].sum():.0f}% of all assemblies:\")\n",
    "top_k_phylum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of species with no phylum information: 446 (2.3%)\n"
     ]
    }
   ],
   "source": [
    "no_phylum_count = len(assemblies[assemblies['phylum'].isnull()])\n",
    "no_phylum_percent = 100 * no_phylum_count / len(assemblies)\n",
    "print(f'Number of species with no phylum information: {no_phylum_count:,} ({no_phylum_percent:.1f}%)')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of distinct phylums (excluding N/A): 166\n"
     ]
    }
   ],
   "source": [
    "n_distinct_phylums = len(assemblies[assemblies['phylum'].notnull()]['phylum'].unique())\n",
    "print(f'Number of distinct phylums (excluding N/A): {n_distinct_phylums:,}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assembly level"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>percent</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>assembly_level</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Scaffold</th>\n",
       "      <td>10509</td>\n",
       "      <td>53.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Contig</th>\n",
       "      <td>6361</td>\n",
       "      <td>32.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Complete Genome</th>\n",
       "      <td>2592</td>\n",
       "      <td>13.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Chromosome</th>\n",
       "      <td>279</td>\n",
       "      <td>1.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 count  percent\n",
       "assembly_level                 \n",
       "Scaffold         10509     53.2\n",
       "Contig            6361     32.2\n",
       "Complete Genome   2592     13.1\n",
       "Chromosome         279      1.4"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly_levels = assemblies[['assembly_level', 'taxid']].groupby('assembly_level').count()\n",
    "assembly_levels.columns = ['count']\n",
    "assembly_levels['percent'] = assembly_levels['count'].apply(lambda c: np.round(100 * c / len(assemblies), 1))\n",
    "assembly_levels.sort_values('count', ascending=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Can I open gzipped files from python?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "# assembly = 'GCA_900110915.1'  ## Royal jelly?\n",
    "assembly = 'GCA_000186885.1'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "taxid                                         577650\n",
       "species_taxid                                    894\n",
       "organism_name     Desulfobulbus propionicus DSM 2032\n",
       "domain                                      Bacteria\n",
       "phylum                                Proteobacteria\n",
       "class                            Deltaproteobacteria\n",
       "order                              Desulfobacterales\n",
       "family                              Desulfobulbaceae\n",
       "genus                                  Desulfobulbus\n",
       "species                    Desulfobulbus propionicus\n",
       "strain            Desulfobulbus propionicus DSM 2032\n",
       "assembly_level                       Complete Genome\n",
       "Name: GCA_000186885.1, dtype: object"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assemblies.loc[assembly]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Pfam records: 7,209\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW16929.1</td>\n",
       "      <td>MreC</td>\n",
       "      <td>PF04085.14</td>\n",
       "      <td>rod-shape-determining-protein-MreC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW18252.1</td>\n",
       "      <td>MrpF_PhaF</td>\n",
       "      <td>PF04066.13</td>\n",
       "      <td>multisubunit-potassium/proton-antiporter,-PhaF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW19265.1</td>\n",
       "      <td>Mrr_cat</td>\n",
       "      <td>PF04471.12</td>\n",
       "      <td>restriction-endonuclease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW19265.1</td>\n",
       "      <td>Mrr_cat_2</td>\n",
       "      <td>PF13156.6</td>\n",
       "      <td>restriction-endonuclease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW19265.1</td>\n",
       "      <td>Mrr_N</td>\n",
       "      <td>PF14338.6</td>\n",
       "      <td>restriction-endonuclease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly_accession  protein_id      query    query_id  \\\n",
       "0    GCA_000186885.1  ADW16929.1       MreC  PF04085.14   \n",
       "1    GCA_000186885.1  ADW18252.1  MrpF_PhaF  PF04066.13   \n",
       "2    GCA_000186885.1  ADW19265.1    Mrr_cat  PF04471.12   \n",
       "3    GCA_000186885.1  ADW19265.1  Mrr_cat_2   PF13156.6   \n",
       "4    GCA_000186885.1  ADW19265.1      Mrr_N   PF14338.6   \n",
       "\n",
       "                                               label  \n",
       "0                 rod-shape-determining-protein-MreC  \n",
       "1  multisubunit-potassium/proton-antiporter,-PhaF...  \n",
       "2                           restriction-endonuclease  \n",
       "3                           restriction-endonuclease  \n",
       "4                           restriction-endonuclease  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfam_path = f'/Users/srom/workspace/nbias/data/sequences/{assembly}/{assembly}_pfam.csv.gz'\n",
    "pfam_df = pd.read_csv(pfam_path)\n",
    "print(f'Number of Pfam records: {len(pfam_df):,}')\n",
    "pfam_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TIGR records: 1,165\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>query</th>\n",
       "      <th>query_id</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW17586.1</td>\n",
       "      <td>TIGR00001</td>\n",
       "      <td>TIGR00001</td>\n",
       "      <td>LSU-ribosomal-protein-L35P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW18531.1</td>\n",
       "      <td>TIGR00002</td>\n",
       "      <td>TIGR00002</td>\n",
       "      <td>SSU-ribosomal-protein-S16P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW18833.1</td>\n",
       "      <td>TIGR00003</td>\n",
       "      <td>TIGR00003</td>\n",
       "      <td>copper-translocating-P-type-ATPase</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW16906.1</td>\n",
       "      <td>TIGR00004</td>\n",
       "      <td>TIGR00004</td>\n",
       "      <td>endoribonuclease-L-PSP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GCA_000186885.1</td>\n",
       "      <td>ADW18617.1</td>\n",
       "      <td>TIGR00005</td>\n",
       "      <td>TIGR00005</td>\n",
       "      <td>pseudouridine-synthase,-RluA-family</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  assembly_accession  protein_id      query   query_id  \\\n",
       "0    GCA_000186885.1  ADW17586.1  TIGR00001  TIGR00001   \n",
       "1    GCA_000186885.1  ADW18531.1  TIGR00002  TIGR00002   \n",
       "2    GCA_000186885.1  ADW18833.1  TIGR00003  TIGR00003   \n",
       "3    GCA_000186885.1  ADW16906.1  TIGR00004  TIGR00004   \n",
       "4    GCA_000186885.1  ADW18617.1  TIGR00005  TIGR00005   \n",
       "\n",
       "                                 label  \n",
       "0           LSU-ribosomal-protein-L35P  \n",
       "1           SSU-ribosomal-protein-S16P  \n",
       "2   copper-translocating-P-type-ATPase  \n",
       "3               endoribonuclease-L-PSP  \n",
       "4  pseudouridine-synthase,-RluA-family  "
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tigr_path = f'/Users/srom/workspace/nbias/data/sequences/{assembly}/{assembly}_tigr.csv.gz'\n",
    "tigr_df = pd.read_csv(tigr_path)\n",
    "print(f'Number of TIGR records: {len(tigr_df):,}')\n",
    "tigr_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>protein_id</th>\n",
       "      <th>distance</th>\n",
       "      <th>probability</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>KB907337.1_1</td>\n",
       "      <td>0.072647</td>\n",
       "      <td>0.927353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>KB907337.1_2</td>\n",
       "      <td>0.082018</td>\n",
       "      <td>0.917982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KB907337.1_3</td>\n",
       "      <td>0.148301</td>\n",
       "      <td>0.851699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KB907337.1_4</td>\n",
       "      <td>0.126539</td>\n",
       "      <td>0.873461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KB907337.1_5</td>\n",
       "      <td>0.167576</td>\n",
       "      <td>0.832424</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     protein_id  distance  probability\n",
       "0  KB907337.1_1  0.072647     0.927353\n",
       "1  KB907337.1_2  0.082018     0.917982\n",
       "2  KB907337.1_3  0.148301     0.851699\n",
       "3  KB907337.1_4  0.126539     0.873461\n",
       "4  KB907337.1_5  0.167576     0.832424"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "assembly = 'GCA_000381985.1'\n",
    "p = os.path.join(\n",
    "    os.getcwd(),\n",
    "    f'data/sequences/{assembly}/{assembly}tri_nucleotide_distance_to_mean.csv',\n",
    ")\n",
    "df = pd.read_csv(p)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtIAAAFsCAYAAADhQl4yAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMywgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/Il7ecAAAACXBIWXMAAAsTAAALEwEAmpwYAAAki0lEQVR4nO3de3BU5f3H8U+WJLsJSUjEXcOAdcAgsaM1QBDLRYroKE0tatNWisq0CnSkIITILYqA3FSwSskfjdJI0Yp1vFRRR7SMWi1eiGIrSDRpMi12TJYEJxeW3HZ/fzhs2V8um/Mk2c0h79dMZtznOZvzTb7s+uHwnGdjAoFAQAAAAAAscUS7AAAAAMCOCNIAAACAAYI0AAAAYIAgDQAAABggSAMAAAAGCNIAAACAAYI0AAAAYCA22gV014kTjfL72fK6vxs6NEk1NQ3RLgMG6J290T/7onf2Rv/s7cz+ORwxSksbbOn5tgnSfn+AIG0T9Mm+6J290T/7onf2Rv/srSf9Y2kHAAAAYIAgDQAAABggSAMAAAAGCNIAAACAAYI0AAAAYIAgDQAAABggSAMAAAAGCNIAAACAAYI0AAAAYIAgDQAAABggSAMAAAAGCNIAAACAAYI0AAAAYCA22gWgfxmckqBEZ/s/FiebWtVY54tCRQAAAP1Tt4L0f/7zH23YsEEffvihBg8erFmzZmnp0qWKjY1VfX291q5dq7feekuJiYn61a9+pV/+8pfB54abR/+S6IyVI//lduP+rderMQr1AAAA9Fdhg3QgENCvf/1rXXjhhXruued0/Phx3X333UpISNBvfvMbFRQUqLq6Wk899ZQqKyu1atUqeTwe5eTkSFLYeQAAAMCOwgZpr9er0aNH67777lNaWppGjRql6667Th9++KG++uor7du3T3v37lVGRoYyMzNVVlamXbt2KScnJ+w8AAAAYFdhbzb0eDx65JFHlJaWJkk6evSo/vrXv2rSpEk6dOiQUlNTlZGRETw+Oztbhw8fVktLS9h5AAAAwK4s7drx4x//WLNmzdKQIUN02223qaqqSh6PJ+QYt9ut1tZWHT9+POw8AAAAYFeWgvSWLVtUXFyshoYG5eXlyefzyel0hhwTHx8vSWpubg47DwAAANiVpe3vvvvd70qSNmzYoFtuuUUTJkxoF4hPP3a5XHK5XF3OWzF0aJKl49H73O7kXj0O/Q+9szf6Z1/0zt7on731pH9hg/Tx48dVUlKia6+9Njg2evRoSVJTU5O8Xm/I8dXV1YqLi1NaWprS09O7nLeipqZBfn/A0nPQ+b7QvuY2JcQPsvS9vN76sMe43cndOg79D72zN/pnX/TO3uifvZ3ZP4cjxvKF27BB+tixY1q8eLHefPNNnX/++ZKkw4cPKzY2VrNmzdKjjz6qiooKjRw5UpJUUlKiSy65RPHx8crKylJNTU2n8+h7Xe0L3dk4AAAAwgu7Rvp73/ueLrvsMq1atUpffPGFDhw4oPvuu0+33Xabhg8frunTp2vFihU6cuSIXn/9de3cuVNz586VpLDzAAAAgF2FvSLtcDi0Y8cObdy4UXPmzFFsbKxuvPFGLV26VNK3NyCuWbNGs2fP1pAhQ7RkyRLNnDkz+Pxw8wAAAIAddetmQ4/Ho0cffbTDudTUVG3fvr3T54abBwAAAOzI0vZ3AAAAAL5FkAYAAAAMEKQBAAAAAwRpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwEBstAuAPZxqaZPbndxu/GRTqxrrfFGoCAAAILoI0ugWV9wgOfJfbjfu33q9GqNQDwAAQLSxtAMAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMBAbLQLQO8YnJKgRCftBAAAiBSS11ki0RkrR/7L7cb9W6+PQjUAAABnP5Z2AAAAAAYI0gAAAIABgjQAAABgoFtB+uuvv9bixYs1ceJETZ48WatXr1ZdXZ0kaffu3RozZkzI14IFC0KeO3/+fI0dO1ZXX321Xnrppb75SQAAAIAICnuzod/v18KFC5Wamqpdu3apublZa9eu1apVq1RYWKgvv/xSN910k/Ly8oLPcTqdwf9euHChhg0bpmeffVYHDx7U6tWrNWLECI0bN65vfiIAAAAgAsIG6dLSUn322Wd699135Xa7JUkFBQWaM2eOGhoaVF5ermuuuSY4d6aPPvpIpaWlKi4uVkpKijIyMvTpp59q9+7dBGkAAADYWtilHcOGDdNjjz0WEpRjYmIUCASCQXrkyJEdPveTTz7RmDFjlJKSEhzLzs7WoUOHel45AAAAEEVhg3RqaqquvPLKkLEnnnhCI0eOVFxcnE6cOKHXXntNM2bM0DXXXKNt27apublZklRVVSWPxxPy3HPPPVdVVVW9+CMAAAAAkWf5A1mKior0xhtvqKioSOXl5ZKk5ORkFRYWqrKyUhs3blRdXZ3WrVsnn88Xsl5akuLj49XW1qbW1lbFxnb/9EOHJlktFRHidid3+Rj2Qe/sjf7ZF72zN/pnbz3pn6UgXVhYqO3bt2vNmjWaOnWqJOn9999XWlqaJCkzM1OStGzZMhUUFMjlcumbb74J+R7Nzc2Ki4uzFKIlqaamQX5/wNJzBpJovoi93vqQOs58DPugd/ZG/+yL3tkb/bO3M/vncMRYvnDb7X2kN23apN/97ndau3at5syZExw/HaJPy8jIUGtrq2pra5Weni6v1xsy7/V62y33AAAAAOymW0F6x44devLJJ7VlyxbNnj07OL5nzx7NmDFDfr8/OHbkyBElJSXJ4/EoKytLpaWlamhoCM6XlJRo7NixvfgjAAAAAJEXNkiXlpaqsLBQt99+uyZPniyv1xv8mjRpkmpra7V+/XpVVlZq//79evDBBzVv3jw5HA5lZ2dr1KhRys/P1xdffKFnnnlGe/fu1a233hqJnw0AAADoM2EXKu/bt09+v19FRUUqKioKmXv11Vf1+OOP66GHHtINN9yglJQUzZ49O/jJhg6HQzt27NC9996r3NxcnXfeedq8ebOysrL65IcBAAAAIiVskF60aJEWLVrU5TF79uzpdG7EiBEqLi62XhkAAADQj3X7ZkMAAAAA/0OQBgAAAAwQpAEAAAADlj/ZENE1OCVBiU7aBgAAEG0kMptJdMbKkf9yu3H/1uujUA0AAMDAxdIOAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwAC7dqBHTrW0ye1ODhlzu5N1sqlVjXW+KFUFAADQ9wjS6BFX3KBOt+NrjEI9AAAAkcLSDgAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAMEaQAAAMAAQRoAAAAw0K0g/fXXX2vx4sWaOHGiJk+erNWrV6uurk6SVF9fr2XLlmn8+PGaOnWqiouLQ54bbh4AAACwo9hwB/j9fi1cuFCpqanatWuXmpubtXbtWq1atUqFhYUqKChQdXW1nnrqKVVWVmrVqlXyeDzKycmRpLDzODudammT253cbvxkU6sa63xRqAgAAKB3hQ3SpaWl+uyzz/Tuu+/K7XZL+jYcz5kzR1999ZX27dunvXv3KiMjQ5mZmSorK9OuXbuUk5MTdh5nL1fcIDnyX2437t96vRqjUA8AAEBvC7u0Y9iwYXrssceCIVqSYmJiFAgEdPDgQaWmpiojIyM4l52drcOHD6ulpUWHDh3qch4AAACwq7BBOjU1VVdeeWXI2BNPPKGRI0eqpqZGHo8nZM7tdqu1tVXHjx9XVVVVl/MAAACAXVnetaOoqEhvvPGGCgoK5PP55HQ6Q+bj4+MlSc3NzWHnAQAAALsKu0b6TIWFhdq+fbvWrFmjqVOn6osvvmgXiE8/drlccrlcXc5bMXRokqXj0X91dBMi+gd6Y2/0z77onb3RP3vrSf+6HaQ3bdqkP/7xj1q7dq1mz54tSUpPT5fX6w05rrq6WnFxcUpLSws7b0VNTYP8/oCl55yNzoYXq9dbH+0S0AG3O5ne2Bj9sy96Z2/0z97O7J/DEWP5wm23lnbs2LFDTz75pLZs2RIM0ZKUlZWlmpoaVVRUBMdKSkp0ySWXKD4+Puw8AAAAYFdhg3RpaakKCwt1++23a/LkyfJ6vcGv9PR0TZ8+XStWrNCRI0f0+uuva+fOnZo7d64kafjw4V3OAwAAAHYVdmnHvn375Pf7VVRUpKKiopC5V199VVu2bNGaNWs0e/ZsDRkyREuWLNHMmTODx4SbBwAAAOwobJBetGiRFi1a1OUx27dv73QuNTW1y3kAAADAjixvfwcAAACAIA0AAAAYIUgDAAAABgjSAAAAgAGCNAAAAGCAIA0AAAAYIEgDAAAABgjSAAAAgAGCNAAAAGCAIA0AAAAYIEgDAAAABgjSAAAAgIHYaBeAgeVUS5vc7uR24yebWtVY54tCRQAAAGYI0ogoV9wgOfJfbjfu33q9GqNQDwAAgCmWdgAAAAAGCNIAAACAAYI0AAAAYIA10v3Q4JQEJTppDQAAQH9GWuuHEp2xHd6QJ317Ux4AAACij6UdAAAAgAGCNAAAAGCAIA0AAAAYIEgDAAAABgjSAAAAgAGCNAAAAGCAIA0AAAAYIEgDAAAABgjSAAAAgAGCNAAAAGCAIA0AAAAYIEgDAAAABgjSAAAAgAGCNAAAAGCAIA0AAAAYIEgDAAAABgjSAAAAgAGCNAAAAGCAIA0AAAAYIEgDAAAABiwF6aamJuXk5Oidd94Jju3evVtjxowJ+VqwYEFw/uuvv9b8+fM1duxYXX311XrppZd6r3oAAAAgSmK7e6DP59PSpUtVVlYWMv7ll1/qpptuUl5eXnDM6XQG/3vhwoUaNmyYnn32WR08eFCrV6/WiBEjNG7cuF4oHwAAAIiObgXpzz77TCtWrFBcXFy7ufLycl1zzTVyu93t5j766COVlpaquLhYKSkpysjI0Keffqrdu3cTpBHiVEub3O7kduMnm1rVWOeLQkUAAABd61aQPnDggKZNm6bFixfrsssuC5krLy/X/PnzO3zeJ598ojFjxiglJSU4lp2drR07dvSgZJyNXHGD5Mh/ud24f+v1aoxCPQAAAOF0K0jPmzevw/GamhqdOHFCr732mtavXy+Hw6HrrrtOixYtUnx8vKqqquTxeEKec+6556qqqqrnlQMAAABR1O010h0pLy+XJCUnJ6uwsFCVlZXauHGj6urqtG7dOvl8vpD10pIUHx+vtrY2tba2Kja2R6cHAAAAoqZHSfbyyy/X+++/r7S0NElSZmamJGnZsmUqKCiQy+XSN998E/Kc5uZmxcXFWQ7RQ4cm9aRU2FhHa6fRN/hd2xv9sy96Z2/0z9560r8eXxI+HaJPy8jIUGtrq2pra5Wenq5//vOfIfNer7fdco/uqKlpkN8f6FGtdsELMpTXWx/tEgYEtzuZ37WN0T/7onf2Rv/s7cz+ORwxli/c9ugDWfbs2aMZM2bI7/cHx44cOaKkpCR5PB5lZWWptLRUDQ0NwfmSkhKNHTu2J6cFAAAAoq5HQXrKlCmqra3V+vXrVVlZqf379+vBBx/UvHnz5HA4lJ2drVGjRik/P19ffPGFnnnmGe3du1e33nprb9UPAAAAREWPgvSIESP0+OOP6+jRo7rhhhu0du1azZ49O/jJhg6HQzt27FBTU5Nyc3P1+OOPa/PmzcrKyuqN2gEAAICosbxGurS0NOTx+PHjtWfPnk6PHzFihIqLi61XBgAAAPRjPboiDQAAAAxUBGkAAADAAEEaAAAAMECQBgAAAAwQpAEAAAADBGkAAADAQI8/IhzmBqckKNFJCwAAAOyIFBdFic5YOfJfbjfu33p9FKoBAACAFSztAAAAAAwQpAEAAAADBGkAAADAAEEaAAAAMECQBgAAAAwQpAEAAAADbH+Hfu1US5vc7uR24yebWtVY54tCRQAAAN8iSKNfc8UN6nSv7cYo1AMAAHAaSzsAAAAAAwRpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAA7HRLgAwcaqlTW53codzJ5ta1Vjni3BFAABgoCFIw5ZccYPkyH+5wzn/1uvVGOF6AADAwMPSDgAAAMAAQRoAAAAwQJAGAAAADBCkAQAAAAOWgnRTU5NycnL0zjvvBMfq6+u1bNkyjR8/XlOnTlVxcXHIc8LNAwAAAHbU7V07fD6fli5dqrKyspDxgoICVVdX66mnnlJlZaVWrVolj8ejnJycbs0DAAAAdtStIP3ZZ59pxYoViouLCxn/6quvtG/fPu3du1cZGRnKzMxUWVmZdu3apZycnLDzAAAAgF11a2nHgQMHNG3aNO3Zsydk/NChQ0pNTVVGRkZwLDs7W4cPH1ZLS0vYeQAAAMCuunVFet68eR2OV1VVyePxhIy53W61trbq+PHjYeeHDRtmWDYAAAAQXT36ZEOfzyen0xkyFh8fL0lqbm4OO2/F0KFJPagUA01nHx+OrvF7szf6Z1/0zt7on731pH89CtIul6tdID792OVyhZ23oqamQX5/oAfV9j+88PqO11sf7RJsx+1O5vdmY/TPvuidvdE/ezuzfw5HjOULtz3aRzo9PV1erzdkrLq6WnFxcUpLSws7DwAAANhVj4J0VlaWampqVFFRERwrKSnRJZdcovj4+LDzAAAAgF31KEgPHz5c06dP14oVK3TkyBG9/vrr2rlzp+bOnduteQAAAMCuerRGWpK2bNmiNWvWaPbs2RoyZIiWLFmimTNndnseAAAAsCPLQbq0tDTkcWpqqrZv397p8eHmgd52qqWtwxs5Tza1qrHOF4WKAADA2ajHV6SB/sYVN0iO/Jfbjfu3Xq/GKNQDAADOTj1aIw0AAAAMVARpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwABBGgAAADDAPtIY8AanJCjR2f6lwAe4AACArhCkMeAlOmP5ABcAAGAZQRoDRmcfHQ4AAGCCII0Bo6uPDgcAALCKmw0BAAAAAwRpAAAAwABBGgAAADDAGukI6Gx7NQAAANgX6S4CutpeDQAAAPbE0g4AAADAAEEaAAAAMECQBgAAAAwQpAEAAAADBGkAAADAAEEaAAAAMECQBgAAAAwQpAEAAAADBGkAAADAAEEaAAAAMECQBgAAAAwQpAEAAAADBGkAAADAAEEaAAAAMECQBgAAAAwQpAEAAAADBGkAAADAAEEaAAAAMECQBgAAAAzERrsAwG4GpyQo0dn+pXOyqVWNdb4oVAQAAKKBIA1YlOiMlSP/5Xbj/q3XqzEK9QAAgOjolaUd+/fv15gxY0K+fvSjH0mS6uvrtWzZMo0fP15Tp05VcXFxb5wSAAAAiKpeuSJdVlam73//+3rooYf+941jv/3WBQUFqq6u1lNPPaXKykqtWrVKHo9HOTk5vXFqoM+cammT250c7TIAAEA/1WtBevTo0XK73SHjX331lfbt26e9e/cqIyNDmZmZKisr065duwjS6PdccYM6XcIBAADQK0s7ysrKNHLkyHbjhw4dUmpqqjIyMoJj2dnZOnz4sFpaWnrj1AAAAEBU9DhIBwIBVVRU6KOPPtLMmTM1ffp0rVmzRvX19aqqqpLH4wk53u12q7W1VcePH+/pqQEAAICo6XGQ/u9//6uTJ08qJiZG27Zt07p163Tw4EHl5eXJ5/PJ6XSGHB8fHy9Jam5u7umpAQAAgKjp8Rrp4cOH64MPPtCQIUMUExMjSUpLS1Nubq4uv/zydoH59GOXy2XpPEOHJvW0VKDP2f3mRLvXP9DRP/uid/ZG/+ytJ/3rlZsNU1NTQx6fXhPd0tIir9cbMlddXa24uDilpaVZOkdNTYP8/kCP6owWXmADh9dbH+0SjLndybauf6Cjf/ZF7+yN/tnbmf1zOGIsX7jt8dKOt99+WxMmTFBDQ0Nw7MiRI3I4HJo1a5ZqampUUVERnCspKdEll1wSXOIBAAAA2FGPg/S4ceOUkJCglStXqry8XB988IEKCgqUm5ur4cOHa/r06VqxYoWOHDmi119/XTt37tTcuXN7o3YAAAAganocpJOTk7Vz5075fD797Gc/01133aUpU6bo3nvvlSRt2bJF6enpmj17tjZu3KglS5Zo5syZPS4cAAAAiKZeWSM9evRo7dy5s8O51NRUbd++vTdOAwAAAPQbvfKBLAAAAMBAQ5AGAAAADPTK0g4A0qmWtg63OjzZ1KrGOl8UKgIAAH2JIA30ElfcIDnyX2437t96vRqjUA8AAOhbBOleNDglQYlOfqUAAAADAamvFyU6Yzu9IgkAAICzCzcbAgAAAAa4Ig30MW5CBADg7ESQBvoYNyECAHB2IkgD/UxnN61yBRsAgP6FIA30M13dtMoVbAAA+g9uNgQAAAAMEKQBAAAAAwRpAAAAwABBGgAAADBAkAYAAAAMEKQBAAAAAwRpAAAAwAD7SANR0tlHhwMAAHsgSANR0tVHhwMAgP6PpR0AAACAAa5IAzY3OCVBic6OX8onm1rVWOeLcEUAAAwMBGnA5hKdsR0uEZG+XSbSGOF6AAAYKFjaAQAAABjgijQwAHW2HGRwSgJLQQAA6CaCNDAAdbYchKUgAAB0H0EasAn2nQYAoH8hSAM2wb7TAAD0L9xsCAAAABggSAMAAAAGWNoBwFhnu3/wQTAAgIGAIA3AGLt/AAAGMoI0cBazutNHZ8dzhRkAgPYI0sBZzOpOH10dzxVmAABCEaQBhMUe1gAAtEeQBhAWe1gDANAe298BAAAABrgibaCzLb8AAAAwcJAGDXS15ReAvt/9g/2rAQD9QUSCdEtLizZv3qxXXnlFkvTTn/5UeXl5cjhYWQKcjTpbU31y8w97JWCzfzUAoD+ISJB++OGH9d5776moqEiNjY1avny5UlJSNH/+/EicHkA/YTVg+5rblBA/qE9r4uo2AMBUnwfppqYmPf3003r00Ud12WWXSZKWLVumbdu2ad68eYqJienrEoyxFhqIjK52BemtZVRdvZ6tXN22Grw7O76zvyR0Nn42B3v+MgPArvo8JX7++efy+XzKzs4OjmVnZ8vr9erYsWM6//zz+7oEY6yFBuwl3H7XVl7PXX0vK8G7q/cRK+O9tSymq79QWP3LQG+em6U6AOyoz4N0VVWVkpKSNHjw4OCY2+0OznU3SDsc0blyfUFaQr8a74819bfx/lgTv4vIjLviBmnkxjc7PL6i4Ope+V6dfZ+mLoJ3X/5sR5dPD573zPOfam6Tq5NlMZ39js78Xt15TkXB1fJZeG9OdMZa+p1KHb/3Jya5lNDRVf6mVp1sONXtevqTaP0/Dr2D/pnrD6/n0/0z6WNMIBAI9HZBZ3rxxRf14IMP6u9//3twzO/36+KLL1ZxcbEmTZrUl6cHAAAA+kSfb5vhcrnU3NwcMnb6scvl6uvTAwAAAH2iz4N0enq66uvr5fP9bx2d1+uVJJ133nl9fXoAAACgT/R5kM7MzFRCQoJKSkqCYwcPHpTH49Hw4cP7+vQAAABAn4jI0o7c3FytX79eH3/8sQ4cOKBt27Zp7ty5fX1qAAAAoM/0+c2G0rd7SW/YsEGvvPKKnE6ncnNzlZeX16/3kAYAAAC6EpEgDQAAAJxt+nxpBwAAAHA2IkgDAAAABgjSAAAAgAGCNLqtpaVF69ev18SJEzVx4kRt3bpVfr+/y+ecOHFCkyZNUnl5eYSqRGes9K+srEy33367srOzNW3aND3wwANqamqKcMU4k5X+ff7557rllluUlZWlq666So899liEq8WZTN47JemOO+7Q0qVLI1AhumKlf7t379aYMWNCvhYsWBDhinGald41Njbqnnvu0YQJE3TFFVfo/vvvb/eBgh1p/+HmQCcefvhhvffeeyoqKlJjY6OWL1+ulJQUzZ8/v8Pja2trtWDBAtXU1ES4UnSku/1rbGzUvHnzNGHCBD3zzDPyer0qKChQW1ubVq9eHaXq0d3+NTQ06Pbbb9d1112nTZs2qaysTPn5+Tr33HN14403Rqn6gc3qe6ckvfDCC/rb3/6mH/7whxGsFB2x0r8vv/xSN910k/Ly8oJjTqczkuXiDFZ6t3LlSlVUVOgPf/iDfD6f8vPzNWTIEC1evLjrkwSAbjh16lTgsssuC7z11lvBseeffz4wefLkgN/vb3f8u+++G7jyyisDs2bNClx00UWBsrKySJaL/8dK/954443A+PHjA01NTcGxv/zlL4GJEydGrF6EstK/0tLSQF5eXqC1tTU4tnDhwsDKlSsjVi/+x+p7ZyAQCFRXVwcmT54cyM3NDSxZsiRSpaIDVvv3i1/8IlBcXBzBCtEZK70rKysLXHTRRYHPP/88OPb0008H7rjjjrDnYWkHuuXzzz+Xz+dTdnZ2cCw7O1ter1fHjh1rd/xbb72lW265Rdu3b49kmeiElf5deumlKiwsVHx8fHAsJiZGDQ0NCrBbZlRY6d9FF12kbdu2adCgQQoEAjp48KA++ugjTZo0KdJlQ9bfOyVp3bp1uvnmm3XhhRdGqkx0wmr/ysvLNXLkyEiWiE5Y6d2BAwc0atQoZWZmBsduvvnmbi2LI0ijW6qqqpSUlKTBgwcHx9xud3Du/ysoKNC8efPkcPBHrD+w0r/zzjtPEydODD5ua2vT7t27NXHiRD5EKUqsvv5OmzBhgubMmaOxY8eyRCBKrPbu1VdfVUVFRZfLPhA5VvpXU1OjEydO6LXXXtOMGTN0zTXXaNu2bd1aZ4veZ6V3//73v/Wd73xHf/rTn3Tttdfqqquu0tatW9XS0hL2PKyRRrf4fL5267xOX7HkTaL/60n/7r//fh09elR//vOf+6w+dM2kf36/X8XFxfr666+1du1abd68Wffcc0+f14pQVnp34sQJbdq0STt27Aj5FyFEj5X+nb6pPjk5WYWFhaqsrNTGjRtVV1endevWRaZgBFnpXWNjo0pKStTS0qIHHnhAtbW1WrdunVpbW7Vy5couz0OQRre4XK52f/BOP3a5XNEoCRaY9K+trU3r1q3Tc889p0cffTTkn7wQWSb9czgcuvTSS3XppZcG70Zfvnw5AS3CrPRuw4YNuu6665SVlRWp8hCGlf5dfvnlev/995WWliZJwffMZcuWqaCggNdehFnp3aBBg9TU1KRHHnlEKSkpkqSTJ0+qoKBAy5cv7/Jf1wnS6Jb09HTV19fL5/MpISFBkuT1eiV9uxQA/ZvV/rW0tCg/P1/79+/X9u3bNWPGjIjWi1BW+nfs2DGVl5dr2rRpwbHRo0erpaVFDQ0NOueccyJXOCz1bu/evXK5XHruueck/e9/+mPHjtUnn3wSwapxmtX3ztMh+rSMjAy1traqtrZW6enpfV8wgqz0zuPxyO12B0O0JI0aNUqnTp1SbW2tzj333E7PwwJWdEtmZqYSEhJUUlISHDt48KA8Ho+GDx8excrQHVb7t2bNGr399tv6/e9/T4juB6z079NPP9Vdd92lxsbG4Njhw4c1dOhQQnQUWOndvn379NJLL+nFF1/Uiy++qKuuukpTpkzRiy++GOGqcZqV/u3Zs0czZswI2af4yJEjSkpKksfjiVjN+JaV3o0bN05VVVUh2/V++eWXSkpKUmpqapfnIUijW1wul3Jzc7V+/Xp9/PHHOnDggLZt26a5c+dKkr755hvV19dHuUp0xkr/3n77bT3//PO6++67NXr0aHm93uAXosNK/37wgx/onHPO0erVq/Wvf/1Lb775ph5++GHdeeed0fwRBiwrvbvgggtCvgYPHqzExERdcMEF0fwRBjQr/ZsyZYpqa2u1fv16VVZWav/+/XrwwQe58T5KrPTuiiuu0MUXX6xly5aptLRUBw4c0MMPP6ybb75ZsbFdL96ICbCfFbqpqalJGzZs0CuvvCKn06nc3Fzl5eUpJiZGt956q4YPH64tW7aEPOfYsWOaMWOGXn31VbZyirLu9m/lypV64YUXOvwe//jHP/hwgSix8vqrqKjQhg0b9PHHHys5OVm33nqr5s2bF+WfYOAyee+Uvv2AiKamJv32t7+NQtU4zUr/SkpK9NBDD+no0aNKSUnRz3/+c915553seBQlVnpXU1Oj+++/X2+//bacTqd+8pOfaMmSJYqLi+vyHARpAAAAwAD/1gAAAAAYIEgDAAAABgjSAAAAgAGCNAAAAGCAIA0AAAAYIEgDAAAABgjSAAAAgAGCNAAAAGCAIA0AAAAY+D9aFp4Qei4OtwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 864x432 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "_, ax = plt.subplots(1, 1, figsize=(12, 6))\n",
    "df['distance'].hist(ax=ax, bins=100);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.9530257799999999, 39, 1.0182767624020888)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "q = df['probability'].quantile(0.99)\n",
    "v = df[df['probability'] > q]\n",
    "q, len(v), 100 * len(v) / len(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
